solarforecastarbiter/metrics/preprocessing.py

```diff
@@ -895,3 +963,143 @@ def _name_pfxobs(current_names, forecast, i=1):
+    drop from analysis. The returned periods do not account for
+    interval label.
+
+    Parameters
+    ----------
+    forecast: solarforecastarbiter.datamodel.Forecast
+    start: pandas.Timestamp
+    end: pandas.Timestamp
+    outages: tuple of solarforecastarbiter.datamodel.TimePeriod
+        List of time ranges to check for forecast issue times.
+
+    Returns
+    -------
+    tuple of solarforecastarbiter.datamodel.TimePeriod
+        Tuple of dicts with start and end keys. Times between these values
+        should not be included in analysis.
+    """
+    # First, determine a list of forecast issue times that include data that
+    # falls within the report
+    issue_times = get_forecast_report_issue_times(forecast, start, end)
+
+    outage_periods = []
+    # For each outage, if a forecast submission/issue_time falls within
+    # the outage. Create start/end bounds for the forecast data to exclude.
+    for outage in outages:
+        outage_submissions = issue_times[
+            (issue_times >= outage.start) & (issue_times <= outage.end)
+        ]
+        for issue_time in outage_submissions:
+            fx_start = issue_time + forecast.lead_time_to_start
+            fx_end = fx_start + forecast.run_length
+            outage_periods.append(datamodel.TimePeriod(
+                start=fx_start,
+                end=fx_end
+            ))
+    return tuple(outage_periods)
+
+
+def remove_outage_periods(
+    outages: Tuple[datamodel.TimePeriod, ...],
+    data: pd.DataFrame,
+    interval_label: str
+) -> Tuple[pd.DataFrame, int]:
+    """Returns a copy of a dataframe with all values within an outage
+    period dropped.
+
+    Parameters
+    ----------
+    outages: tuple of :py:class:`solarforecastarbiter.datamodel.TimePeriod`
+        Tuple of dictionaries with start and end keys. Values should be
+        timestamps denoting the start and end of periods to remove.
+    data: pandas.DataFrame
+        The dataframe to drop outage data from.
+    interval_label: str
+        The interval label to drop.
+
+    Returns
+    -------
+    pandas.DataFrame, int
+        The data DataFrame with outage data dropped, and total
+        number of points removed.
+    """  # NOQA
+    if len(outages) == 0:
+        return data, 0
+    dropped_total = 0
+
+    # Set to the boolean series of outage data on first iteration
+    # of loop below
+    full_outage_index = None
+
+    for outage in outages:
+        if interval_label == "ending":
+            outage_index = (data.index > outage.start) & (
+                data.index <= outage.end)
+        else:
+            outage_index = (data.index >= outage.start) & (
+                data.index < outage.end)
+        if full_outage_index is None:
+            full_outage_index = outage_index
+        else:
+            full_outage_index = full_outage_index | outage_index
+        dropped_total += outage_index.sum()
+    return data[~full_outage_index], dropped_total
```

Comment:
why do we compute `dropped_total += outage_index.sum()` within the loop instead of `full_outage_index.sum()` at the end? I feel like we discussed this but I've forgotten. If there's a reason it should be in a comment